torch_xla/amp/syncfree/sgd.py

```diff
@@ -0,0 +1,199 @@
+            &\hspace{15mm} g_t  \leftarrow  \textbf{b}_t                                         \\
+            &\hspace{5mm}\theta_t \leftarrow \theta_{t-1} - \gamma g_t                    \\[-1.ex]
+            &\rule{110mm}{0.4pt}                                                          \\[-1.ex]
+            &\bf{return} \:  \theta_t                                                     \\[-1.ex]
+            &\rule{110mm}{0.4pt}                                                          \\[-1.ex]
+       \end{aligned}
+
+    Nesterov momentum is based on the formula from
+    `On the importance of initialization and momentum in deep learning`__.
+
+    Args:
+        params (iterable): iterable of parameters to optimize or dicts defining
+            parameter groups
+        lr (float): learning rate
+        momentum (float, optional): momentum factor (default: 0)
+        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+        dampening (float, optional): dampening for momentum (default: 0)
+        nesterov (bool, optional): enables Nesterov momentum (default: False)
+
+    Example:
+        >>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
+        >>> optimizer.zero_grad()
+        >>> loss_fn(model(input), target).backward()
+        >>> optimizer.step(found_inf=torch.tensor(1)) # skip the step
+        >>> optimizer.step(found_inf=torch.tensor(0)) # perform the step
+
+    __ http://www.cs.toronto.edu/%7Ehinton/absps/momentum.pdf
+
+    .. note::
+        The implementation of SGD with Momentum/Nesterov subtly differs from
+        Sutskever et. al. and implementations in some other frameworks.
+
+        Considering the specific case of Momentum, the update can be written as
+
+        .. math::
+            \begin{aligned}
+                v_{t+1} & = \mu * v_{t} + g_{t+1}, \\
+                p_{t+1} & = p_{t} - \text{lr} * v_{t+1},
+            \end{aligned}
+
+        where :math:`p`, :math:`g`, :math:`v` and :math:`\mu` denote the
+        parameters, gradient, velocity, and momentum respectively.
+
+        This is in contrast to Sutskever et. al. and
+        other frameworks which employ an update of the form
+
+        .. math::
+            \begin{aligned}
+                v_{t+1} & = \mu * v_{t} + \text{lr} * g_{t+1}, \\
+                p_{t+1} & = p_{t} - v_{t+1}.
+            \end{aligned}
+
+        The Nesterov version is analogously modified.
+    """
+
+  def __init__(self,
+               params,
+               lr=required,
+               momentum=0,
+               dampening=0,
+               weight_decay=0,
+               nesterov=False):
+    if lr is not required and lr < 0.0:
+      raise ValueError("Invalid learning rate: {}".format(lr))
+    if momentum < 0.0:
+      raise ValueError("Invalid momentum value: {}".format(momentum))
+    if weight_decay < 0.0:
+      raise ValueError("Invalid weight_decay value: {}".format(weight_decay))
+
+    defaults = dict(
```

Comment:
I am less familiar with this part, I am guessing it is mostly the same as the pytorch sgd? The only difference is it calls pt/xla's  `_xla_sgd_optimizer_step` in the end.

Reply:
That's right, the only difference is using `_xla_sgd_optimizer_step` to replace `F.sgd` and adding a `step` counter.
https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py